File: src/services/openlist.ts
import { invoke } from "@tauri-apps/api/core";
import type {
  FileListResponse,
  FileOperationResult,
  ServerTestResult,
  FileInfo,
} from "@/types";

function isTauriAvailable(): boolean {
  return typeof window !== "undefined" && 
    "__TAURI_INTERNALS__" in window;
}

function checkTauriEnvironment(): void {
  if (!isTauriAvailable()) {
    throw new Error("Tauri API 不可用。请在 Tauri 环境中运行此应用，或使用 `npm run tauri dev` 启动开发模式。");
  }
}

function isTokenInvalidatedError(error: unknown): boolean {
  const errorMessage = error instanceof Error ? error.message : String(error);
  return errorMessage.toLowerCase().includes("invalidated") || 
         errorMessage.toLowerCase().includes("token expired") ||
         errorMessage.toLowerCase().includes("unauthorized") ||
         errorMessage.toLowerCase().includes("401");
}

export async function executeWithTokenRefresh<T>(
  operation: () => Promise<T>,
  serverId: string,
  serverUrl: string,
  username: string,
  password: string,
  updateServerToken: (serverId: string, newToken: string) => Promise<void>
): Promise<T> {
  try {
    return await operation();
  } catch (error) {
    if (isTokenInvalidatedError(error) && username && password) {
      console.log(`[OpenList] Token 已失效，正在重新登录服务器: ${serverUrl}`);
      try {
        const newToken = await loginToOpenlist(serverUrl, username, password);
        await updateServerToken(serverId, newToken);
        console.log(`[OpenList] Token 刷新成功，重试操作`);
        return await operation();
      } catch (loginError) {
        console.error(`[OpenList] Token 刷新失败:`, loginError);
        throw loginError;
      }
    }
    throw error;
  }
}

export function validateServerUrl(url: string): { valid: boolean; error?: string; normalizedUrl?: string } {
  if (!url || url.trim() === "") {
    return { valid: false, error: "服务器地址不能为空" };
  }

  let normalizedUrl = url.trim();
  
  if (!normalizedUrl.startsWith("http://") && !normalizedUrl.startsWith("https://")) {
    normalizedUrl = "https://" + normalizedUrl;
  }

  try {
    const urlObj = new URL(normalizedUrl);
    if (urlObj.protocol !== "http:" && urlObj.protocol !== "https:") {
      return { valid: false, error: "仅支持 HTTP 或 HTTPS 协议" };
    }
    if (!urlObj.hostname || urlObj.hostname === "") {
      return { valid: false, error: "服务器地址格式不正确，缺少主机名" };
    }
    return { valid: true, normalizedUrl: normalizedUrl.replace(/\/+$/, "") };
  } catch {
    return { valid: false, error: "服务器地址格式不正确，请使用有效的 URL（如 https://example.com）" };
  }
}

export async function testConnection(url: string, token: string): Promise<ServerTestResult> {
  checkTauriEnvironment();
  return invoke("test_openlist_connection", { url, token });
}

export async function loginToOpenlist(url: string, username: string, password: string, otpCode?: string): Promise<string> {
  checkTauriEnvironment();
  return invoke("login_to_openlist", { url, username, password, otpCode: otpCode || null });
}

export async function listDirectory(url: string, token: string, path: string): Promise<FileListResponse> {
  checkTauriEnvironment();
  return invoke("list_directory", { url, token, path });
}

export async function renameFile(url: string, token: string, dir: string, oldName: string, newName: string): Promise<FileOperationResult> {
  checkTauriEnvironment();
  return invoke("rename_file", { url, token, dir, oldName, newName });
}

export async function deleteFiles(url: string, token: string, dir: string, names: string[]): Promise<FileOperationResult> {
  checkTauriEnvironment();
  return invoke("delete_files", { url, token, dir, names });
}

export async function copyFiles(url: string, token: string, srcDir: string, dstDir: string, names: string[]): Promise<FileOperationResult> {
  checkTauriEnvironment();
  return invoke("copy_files", { url, token, srcDir, dstDir, names });
}

export async function moveFiles(url: string, token: string, srcDir: string, dstDir: string, names: string[]): Promise<FileOperationResult> {
  checkTauriEnvironment();
  return invoke("move_files", { url, token, srcDir, dstDir, names });
}

export async function getFileInfo(url: string, token: string, path: string): Promise<FileInfo> {
  checkTauriEnvironment();
  return invoke("get_file_info", { url, token, path });
}

File: src-tauri/src/commands/openlist.rs
//! OpenList 命令模块
//!
//! 本模块提供与 OpenList 服务器交互的 Tauri 命令。这些命令作为前端与后端服务层
//! 之间的桥梁,处理用户认证、文件浏览、文件操作等请求。
//!
//! 主要功能:
//! - 用户登录认证
//! - 服务器连接测试
//! - 目录浏览和文件列表获取
//! - 文件重命名、删除、复制、移动操作
//! - 文件详情查询

use crate::models::openlist::*;
use crate::services::openlist::OpenListService;

/// 登录到 OpenList 服务器
///
/// 使用提供的用户名和密码进行认证,获取访问令牌。支持可选的 OTP 验证码
/// (用于双因素认证场景)。
///
/// # 参数
///
/// * `url` - OpenList 服务器地址
/// * `username` - 用户名
/// * `password` - 密码
/// * `otp_code` - 可选的 OTP 验证码(用于双因素认证)
///
/// # 返回值
///
/// 成功时返回访问令牌字符串,失败时返回错误信息
#[tauri::command]
pub async fn login_to_openlist(url: String, username: String, password: String, otp_code: Option<String>) -> Result<String, String> {
    let service = OpenListService::new();
    service.login(&url, &username, &password, otp_code).await
}

/// 测试 OpenList 服务器连接
///
/// 验证提供的服务器地址和访问令牌是否有效,并返回服务器测试结果。
///
/// # 参数
///
/// * `url` - OpenList 服务器地址
/// * `token` - 访问令牌
///
/// # 返回值
///
/// 成功时返回 `ServerTestResult` 包含连接状态和服务器信息,失败时返回错误信息
#[tauri::command]
pub async fn test_openlist_connection(url: String, token: String) -> Result<ServerTestResult, String> {
    let service = OpenListService::new();
    service.test_connection(&url, &token).await
}

/// 列出目录内容
///
/// 获取指定路径下的文件和文件夹列表。
///
/// # 参数
///
/// * `url` - OpenList 服务器地址
/// * `token` - 访问令牌
/// * `path` - 要列出的目录路径
///
/// # 返回值
///
/// 成功时返回 `FileListResponse` 包含文件列表,失败时返回错误信息
#[tauri::command]
pub async fn list_directory(url: String, token: String, path: String) -> Result<FileListResponse, String> {
    let service = OpenListService::new();
    service.list_directory(&url, &token, &path).await
}

/// 重命名文件
///
/// 将指定目录中的文件从旧名称重命名为新名称。
///
/// # 参数
///
/// * `url` - OpenList 服务器地址
/// * `token` - 访问令牌
/// * `request` - 重命名请求,包含目录路径、旧名称和新名称
///
/// # 返回值
///
/// 成功时返回 `FileOperationResult` 包含操作结果,失败时返回错误信息
#[tauri::command]
pub async fn rename_file(url: String, token: String, request: RenameRequest) -> Result<FileOperationResult, String> {
    let service = OpenListService::new();
    service.rename(&url, &token, &request.dir, &request.old_name, &request.new_name).await
}

/// 删除文件
///
/// 批量删除指定目录中的一个或多个文件。
///
/// # 参数
///
/// * `url` - OpenList 服务器地址
/// * `token` - 访问令牌
/// * `dir` - 文件所在目录路径
/// * `names` - 要删除的文件名列表
///
/// # 返回值
///
/// 成功时返回 `FileOperationResult` 包含操作结果,失败时返回错误信息
#[tauri::command]
pub async fn delete_files(url: String, token: String, dir: String, names: Vec<String>) -> Result<FileOperationResult, String> {
    let service = OpenListService::new();
    service.delete(&url, &token, &dir, names).await
}

/// 复制文件
///
/// 将一个或多个文件从源目录复制到目标目录。
///
/// # 参数
///
/// * `url` - OpenList 服务器地址
/// * `token` - 访问令牌
/// * `request` - 复制请求,包含源目录、目标目录和文件名列表
///
/// # 返回值
///
/// 成功时返回 `FileOperationResult` 包含操作结果,失败时返回错误信息
#[tauri::command]
pub async fn copy_files(url: String, token: String, request: CopyMoveRequest) -> Result<FileOperationResult, String> {
    let service = OpenListService::new();
    service.copy(&url, &token, &request.src_dir, &request.dst_dir, request.names).await
}

/// 移动文件
///
/// 将一个或多个文件从源目录移动到目标目录。
///
/// # 参数
///
/// * `url` - OpenList 服务器地址
/// * `token` - 访问令牌
/// * `request` - 移动请求,包含源目录、目标目录和文件名列表
///
/// # 返回值
///
/// 成功时返回 `FileOperationResult` 包含操作结果,失败时返回错误信息
#[tauri::command]
pub async fn move_files(url: String, token: String, request: CopyMoveRequest) -> Result<FileOperationResult, String> {
    let service = OpenListService::new();
    service.move_files(&url, &token, &request.src_dir, &request.dst_dir, request.names).await
}

/// 获取文件信息
///
/// 获取指定路径下文件的详细信息,包括文件大小、修改时间、类型等元数据。
///
/// # 参数
///
/// * `url` - OpenList 服务器地址
/// * `token` - 访问令牌
/// * `path` - 文件路径
///
/// # 返回值
///
/// 成功时返回 `FileInfo` 包含文件详细信息,失败时返回错误信息
#[tauri::command]
pub async fn get_file_info(url: String, token: String, path: String) -> Result<FileInfo, String> {
    let service = OpenListService::new();
    service.get_file_info(&url, &token, &path).await
}
